docs/RationaleSymbols.md

```diff
@@ -0,0 +1,79 @@
+# Symbol and Inner Symbol Rationale
+
+This document describes various design points of the major CIRCT dialects 
+relating to the use of symbols and the introduction of inner symbols and 
+related types.  This follows in the spirit of other 
+[MLIR Rationale docs](https://mlir.llvm.org/docs/Rationale/).
+
+## Introduction
+
+Verilog and FIRRTL have, from a software compiler perspective, an unusual 
+number of namable entities which can be refered to non-locally.  These entities 
+have deep nesting in the code structures.  The requirements of dealing with 
+these entities and references entails a more complexity than provided by MLIR's 
+symbols and symbol tables.  Several Circt dialects, therefor, share a common 
+supplimental mechanism called "Inner Symbols" to manage these requirements.  
+Inner Symbols necessarily deviates from MLIR nested symbol tables to enable 
+representation of the behavior of verilog and firrtl.
+
+## Use of MLIR symbols
+
+MLIR symbols are directly used for items in the global scope.  This primarily 
+consists of hw or firrtl modules, though other entities, such as system verilog 
+interfaces and firrtl non-local anchors and bind statements, also share this 
+space.  Modules and instances of them are well suited to mlir symbols.  They 
+are analogous in scoping and structure to functions and call instructions.  The 
+top-level mlir module or firrtl circuit op define a symbol table and all 
+modules contained define symbols, with instances refering by symbol to their 
+instantiated module.
+
+## Inner Symbol
+
+Within a module (firrtl or hw), many entities may exist which can be referenced 
+outside the module.  Operations and ports (and memory ports), need to define 
+symbol-like data to allow forming non-ssa linkage between disparate elements.  
+To accomplish this, an attribute `inner_sym` is created providing a scoped 
+symbol-like name to the element.  An `inner_sym` operation resides in 
+arbitrarily-nested regions of a region that defines an `InnerSymbolTable` and
+an `Symbol` .
```

Comment:
I'm assuming `inner_sym` is the attribute name, not the name of the op interface?
```suggestion
Within a `firrtl` or `hw` module, many entities may exist which can be referenced 
outside the module.  Operations and ports (and memory ports), need to define 
symbol-like data to allow forming non-ssa linkage between disparate elements.  
To accomplish this, an attribute named `inner_sym` is attached providing a scoped 
symbol-like name to the element.  An operation with an `inner_sym` resides in 
arbitrarily-nested regions of a region that defines an `InnerSymbolTable` and
a `Symbol`.
```